File: frontend/src/types/analysis.ts
export type SignalDirection = "buy" | "sell" | "neutral" | "caution" | "info";
export type SignalStrength = "weak" | "medium" | "strong" | "info";
export type AggregateSignal =
  | "strong_buy"
  | "buy"
  | "neutral"
  | "sell"
  | "strong_sell"
  | "caution";
export type Confidence = "low" | "medium" | "high";

export type TechnicalSignalItem = {
  indicator_name: string;
  label: string;
  value: Record<string, number> | number | string | null;
  signal: SignalDirection;
  score: number;
  strength: SignalStrength;
  reason: string;
  timestamp: string | null;
};

export type TechnicalSignalAggregate = {
  instrument_id: number;
  timeframe: string;
  total_score: number;
  signal: AggregateSignal;
  confidence: Confidence;
  bullish_count: number;
  bearish_count: number;
  caution_count: number;
  info_count: number;
  generated_at: string;
};

export type TechnicalSignalResponse = {
  instrument_id: number;
  timeframe: string;
  aggregate: TechnicalSignalAggregate;
  signals: TechnicalSignalItem[];
};

File: backend/app/schemas/analysis.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel

SignalDirection = Literal["buy", "sell", "neutral", "caution", "info"]
SignalStrength = Literal["weak", "medium", "strong", "info"]
AggregateSignal = Literal["strong_buy", "buy", "neutral", "sell", "strong_sell", "caution", "no_data"]
Confidence = Literal["low", "medium", "high"]


class TechnicalSignalItem(BaseModel):
    indicator_name: str
    label: str
    value: dict[str, Any] | float | str | None
    signal: SignalDirection
    score: int
    strength: SignalStrength
    reason: str
    timestamp: datetime | None


class TechnicalSignalAggregate(BaseModel):
    instrument_id: int
    timeframe: str
    total_score: int
    signal: AggregateSignal
    confidence: Confidence
    bullish_count: int
    bearish_count: int
    caution_count: int
    info_count: int
    generated_at: datetime


class TechnicalSignalResponse(BaseModel):
    instrument_id: int
    timeframe: str
    aggregate: TechnicalSignalAggregate
    signals: list[TechnicalSignalItem]
    message: str | None = None
